tests/DCPS/Presentation/main.cpp

```diff
@@ -8,350 +10,584 @@
-      {
-        const Foo foo = {0, 0, 0, 0};
-        if (writer_i->write(foo, DDS::HANDLE_NIL) != DDS::RETCODE_OK)
-        {
-            ACE_ERROR_RETURN((LM_ERROR,
-                              ACE_TEXT("%N:%l main()")
-                              ACE_TEXT(" ERROR: Unable to write sample!\n")), -1);
-        }
+    if (!CORBA::is_nil(subscriber_)) {
+      rc = participant_->delete_subscriber(subscriber_);
+      if (rc != DDS::RETCODE_OK) {
+        ACE_ERROR((LM_ERROR, ACE_TEXT("%N:%l ~Test() ERROR: ")
+          ACE_TEXT("delete subscriber failed: %C\n"),
+          retcode_to_dds_string(rc).c_str()));
       }
-      ACE_OS::sleep(5); // wait for samples to arrive
-
-      // Verify that no samples are yet available
-      {
-        FooSeq foo;
-        DDS::SampleInfoSeq info;
-
-        DDS::ReturnCode_t error;
-        if ((error = reader_i->take(foo,
-                                    info,
-                                    DDS::LENGTH_UNLIMITED,
-                                    DDS::ANY_SAMPLE_STATE,
-                                    DDS::ANY_VIEW_STATE,
-                                    DDS::ANY_INSTANCE_STATE)) != DDS::RETCODE_NO_DATA)
-        {
-          ACE_ERROR_RETURN((LM_ERROR,
-                            ACE_TEXT("%N:%l main()")
-                            ACE_TEXT(" ERROR: Unexpected samples read!\n")), -1);
-        }
+    }
+
+    if (!CORBA::is_nil(publisher_)) {
+      rc = participant_->delete_publisher(publisher_);
+      if (rc != DDS::RETCODE_OK) {
+        ACE_ERROR((LM_ERROR, ACE_TEXT("%N:%l ~Test() ERROR: ")
+          ACE_TEXT("delete publisher failed: %C\n"),
+          retcode_to_dds_string(rc).c_str()));
       }
+    }
+  }
+};
 
-      publisher->end_coherent_changes();
-      ACE_OS::sleep(5); // wait for samples to arrive
-
-      //Verify coherent samples are now available
-      {
-        FooSeq foo;
-        DDS::SampleInfoSeq info;
-
-        DDS::ReturnCode_t error;
-        if ((error = reader_i->take(foo,
-                                    info,
-                                    DDS::LENGTH_UNLIMITED,
-                                    DDS::ANY_SAMPLE_STATE,
-                                    DDS::ANY_VIEW_STATE,
-                                    DDS::ANY_INSTANCE_STATE)) != DDS::RETCODE_OK)
-        {
-          ACE_ERROR_RETURN((LM_ERROR,
-                            ACE_TEXT("%N:%l main()")
-                            ACE_TEXT(" ERROR: take failed %d\n"), error), -1);
-        }
+bool
+coherent_test(DDS::DomainParticipant_var& participant, DDS::Topic_var& topic)
+{
+  DDS::ReturnCode_t rc;
+
+  ACE_DEBUG((LM_DEBUG, ACE_TEXT("Testing coherent_access...\n")));
+
+  DDS::Publisher_var publisher;
+  DDS::Subscriber_var subscriber;
+  DDS::DataReader_var reader1;
+  DDS::DataReader_var reader2;
+  DDS::DataWriter_var writer;
+  Test t(participant, publisher, subscriber, reader1, reader2, writer);
+  if (t.setup_test(topic, true)) {
```

Comment:
Does `setup_test` return `true` if it fails?  Should it?
(Look like this applies to a number of functions in the test)

Reply:
Yes, I think it should be an check for "did this produce an error", since errors are the exceptional case in this test. If it's a matter of returning true vs. false, using false for error means using `!` everywhere to check for an error case.

Reply:
I don't think we've ever used this pattern (but correct me if I'm wrong).  I'm pretty sure we've used the opposite.  As is, I think this is too confusing for someone reading the code.  Using `!` seems like the right thing to do, just like checking for a null pointer.

Reply:
Could've sworn that's what the convention was in QueryCondition test, but I can see it's not. I will concede that's what the pattern is and will change it.

I still think the pattern is wrong though. Most of the time we're not expecting an error, so the pattern should be simpler than the opposite case.